horovod/common/controller.cc

```diff
@@ -0,0 +1,757 @@
+// Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include <atomic>
+#include <map>
+#include <queue>
+#include <set>
+#include <unordered_set>
+
+#include "controller.h"
+#include "global_state.h"
+#include "logging.h"
+#include "operations.h"
+
+namespace horovod {
+namespace common {
+
+Controller::Controller(ResponseCache& response_cache,
+                       TensorQueue& tensor_queue, Timeline& timeline,
+                       ParameterManager& parameter_manager)
+    : stall_inspector_(response_cache), tensor_queue_(tensor_queue),
+      timeline_(timeline), response_cache_(response_cache),
+      parameter_manager_(parameter_manager) {}
+
+// This function performs all the preparation work for workers to agree
```

Comment:
Usually you want to put member function comments in the header file, as that's the first place a user is likely to look.

Reply:
Moved.